framework/master-key-provider-keyring.md

```diff
@@ -0,0 +1,125 @@
+
+For each master key, OnEncrypt MUST do the following with the response from [Encrypt Data Key](master-key-interface.md#encrypt-data-key)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the master key.
+- append a new record to the [keyring trace](structures.md#keyring-trace)
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the master key.
+  - The string field KeyName contains the key id of the master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+### OnDecrypt
+
+If the input [decryption materials](structures.md#decryption-materials) contain a plaintext data key,
+OnDecrypt MUST immediately return the unmodified decryption materials.
+
+Otherwise, OnDecrypt MUST attempt to decrypt the [encrypted data keys](structures.md#encrypted-data-key)
+in the input using [Decrypt Data Key](master-key-provider-interface.md#decrypt-data-key)
+on the [master key provider](master-key-provider-interface.md).
+
+The input to [DecryptDataKey](master-key-provider-interface.md#decrypt-data-key) MUST contain the following:
+ - [Algorithm Suite](structures.md#algorithm-suite)  from the input [decryption materials](structures.md#decryption-materials)
+ - [Encryption Context](structures.md#encryption-context) from the input [decryption materials](structures.md#decryption-materials)
+ - [Encrypted Data Keys](structures.md#plaintext-data-key) from the input
+
+If OnDecrypt fails to successfully decrypt the [encrypted data keys](structures.md#encrypted-data-key),
+then OnDecrypt MUST output the unmodified input [decryption materials](structures.md#decryption-materials).
+
+Otherwise, OnDecrypt MUST do the following with the response from [Decrypt Data Key](master-key-provider-interface.md#decrypt-data-key)
+- set the plaintext data key on the [decryption materials](structures.md#decryption-materials)
+- append a new record to the [keyring trace](structures.md#keyring-trace)
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the master key that was used to decrypt the data key.
+  - The string field KeyName contains the key id of the master key that was used to decrypt the data key.
```

Comment:
Ditto type and names being Java-specific.

Reply:
Updated